feat(customTemplate): create flipping card

simboonlong opened the pull request:
add css-loader, node-sass, sass-loader, style-loader
add scss loading into storybook webpack config
flip card on click

simboonlong committed:
feat(customTemplate): create flipping card

add css-loader, node-sass, sass-loader, style-loader
add scss loading into storybook webpack config
flip card on click

Nebulis commented on .storybook/webpack.config.js:
```diff
@@ -19,6 +21,14 @@ module.exports = ({ config }) => {
   });
+
+  // loader for scss files
+  config.module.rules.push({
```
can we not use sass ?
- why do we need sass ?
- node-sass is pretty annoying (it's compiled on every machine ..)
- this project already uses some css lib (let's avoid to use too many things)

simboonlong committed:
remove css-loader, node-sass, sass-loader, style-loader
remove scss loading from storybook webpack config
remove scss folder
use css-in-js

simboonlong committed:
refactor to use emotion css

sebastianquek commented on src/templates/customTemplate/customTemplate.tsx:
```diff
@@ -48,59 +97,69 @@ const Ratings: FunctionComponent<RatingsProps> = ({ ratings }: { ratings: Rating
+    } else {
+      setFlip(true);
+    }
+  };
```
We can do `setFlip(prevFlip => !prevFlip)` instead

sebastianquek reviewed (commented):
## General comments
- Good use of CSS transforms!
- The flipping interaction is not obvious at the moment, shall we add a button to indicate that the card can be flipped?

## Technical comments
- I noticed you used emotion's `css(...)`, inline css (`<div css={...}>`) and also `className` (which mostly uses bootstrap's classes). The styles are therefore scattered across several places which makes it quite hard to reason about. It's also challenging to logically link the styles with its corresponding use cases.
- I suggest we minimise the approaches for styling. I prefer using Emotion's `styled` approach (https://emotion.sh/docs/styled). This keeps the markup relatively clean and we treat each styled component like any other react component that accepts props.
```
const FlipCardFace = styled.div<{isFront: boolean}>`
  transform: rotateY(${({isFront}) => isFront ? '0deg' : '180deg'});
  ...
`;
...
<FlipCardFace isFront={true}>
```
- The styles are therefore more focused on the component level so we can look through the styles and understand quickly its use cases. This is opposed to specific properties that doesn't express where it's to be used. (e.g. `borderLeft`, though if we want to share styles, we can still fallback to this). 

Let me know what you think!

simboonlong committed:
use styled approach
refactor divs into components

simboonlong committed:
add expiring, expired label status

simboonlong committed:
check date expiring for 3months before

simboonlong committed:
rectify linter errors

simboonlong committed:
add arrow cta to flip

simboonlong committed:
testing emotion keyframe animation

simboonlong committed:
shift arrow button outside of transformed card to bypass card being "blank" bug

simboonlong committed:
remove unused animation
return null for date check condition
use button instead of div for semantic purposes

sebastianquek committed:
test: add LabelStatus tests

simboonlong committed:
import svg html directly so no need to use webpack loaders

simboonlong committed:
refactor LabelStatus out as a component
refactor const to separate file to reuse
remove declaration of svg as module, no longer needed

simboonlong committed:
remove svg images

sebastianquek reviewed (approved):
👍
